pkg/common/cliprinter/cliprinter.go

```diff
@@ -0,0 +1,131 @@
+package cliprinter
+
+import (
+	"io"
+	"os"
+	"sync"
+
+	"github.com/spiffe/spire/pkg/common/cliprinter/errorjson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/errorpretty"
+	"github.com/spiffe/spire/pkg/common/cliprinter/protojson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/protopretty"
+	"github.com/spiffe/spire/pkg/common/cliprinter/structjson"
+	"github.com/spiffe/spire/pkg/common/cliprinter/structpretty"
+	"google.golang.org/protobuf/proto"
+)
+
+type Printer interface {
+	PrintError(error)
+	PrintProto(...proto.Message)
+	PrintStruct(...interface{})
+}
+
+type DefaultPrinter struct {
+	format formatType
+	mtx    sync.RWMutex
+
+	stdout io.Writer
+	stderr io.Writer
+
+	customPrettyFunc func(...interface{})
+}
+
+// NewDefaultPrinter returns a new printer configured with
+// the default formatter
+func NewDefaultPrinter() *DefaultPrinter {
+	return newPrinter(pretty)
+}
+
+// NewPrinter returns a new printer and configures it to use
+// the specified formatter
+func NewPrinter(format string) (*DefaultPrinter, error) {
+	f, err := strToFormatType(format)
```

Comment:
Just curious, why not have the `NewPrinter()` func accept the `formatType` enum and make it public?

Reply:
The intention here was to keep the consumer requirement as low as possible, and also to provide a stable experience for all consumers/users.

Exporting the enum type would force consumers to switch on them based on the provided flag values which could introduce variability (e.g. json vs JSON vs JSon). Alternatively, we could expose the functions to convert the string to format type, but then that would require the caller to call that function too and handle potential errors. Just felt easier to package it all up here (well .. either here or in flag.go)